platform/default/src/mbgl/map/map_snapshotter.cpp

```diff
@@ -4,218 +4,302 @@
+        : renderer(std::make_unique<util::Thread<SnapshotterRenderer>>(
+              "Snapshotter", size, pixelRatio, std::move(localFontFamily))) {}
+
+    ~SnapshotterRendererFrontend() = default;
 
-    if (cameraOptions) {
-        map.jumpTo(*cameraOptions);
+    void reset() override { renderer->actor().invoke(&SnapshotterRenderer::reset); }
+
+    void setObserver(RendererObserver& observer) override {
+        renderer->actor().invoke(&SnapshotterRenderer::setObserver,
+                                 std::make_unique<ForwardingRendererObserver>(observer));
     }
 
-    // Set region, if specified
-    if (region) {
-        this->setRegion(*region);
+    void update(std::shared_ptr<UpdateParameters> parameters) override {
+        updateParameters = std::move(parameters);
+        renderer->actor().invoke(&SnapshotterRenderer::update, updateParameters);
     }
-}
 
-void MapSnapshotter::Impl::snapshot(ActorRef<MapSnapshotter::Callback> callback) {
-    map.renderStill([this, callback = std::move(callback)] (std::exception_ptr error) {
-
-        // Create lambda that captures the current transform state
-        // and can be used to translate for geographic to screen
-        // coordinates
-        assert (frontend.getTransformState());
-        PointForFn pointForFn { [=, center = *map.getCameraOptions().center, transformState = *frontend.getTransformState()] (const LatLng& latLng) {
-            LatLng unwrappedLatLng = latLng.wrapped();
-            unwrappedLatLng.unwrapForShortestPath(center);
-            Transform transform { transformState };
-            return transform.latLngToScreenCoordinate(unwrappedLatLng);
-        }};
-
-        // Create lambda that captures the current transform state
-        // and can be used to translate for geographic to screen
-        // coordinates
-        assert (frontend.getTransformState());
-        LatLngForFn latLngForFn { [=, transformState = *frontend.getTransformState()] (const ScreenCoordinate& screenCoordinate) {
-            Transform transform { transformState };
-            return transform.screenCoordinateToLatLng(screenCoordinate);
-        }};
-
-        // Collect all source attributions
-        std::vector<std::string> attributions;
-        for (auto source : map.getStyle().getSources()) {
-            auto attribution = source->getAttribution();
-            if (attribution) {
-                attributions.push_back(*attribution);
-            }
-        }
+    void setSize(Size size) { renderer->actor().invoke(&SnapshotterRenderer::setSize, size); }
 
-        // Invoke callback
-        callback.invoke(
-                &MapSnapshotter::Callback::operator(),
-                error,
-                error ? PremultipliedImage() : frontend.readStillImage(),
-                std::move(attributions),
-                std::move(pointForFn),
-                std::move(latLngForFn)
-        );
-    });
-}
+    optional<TransformState> getTransformState() const {
```

Comment:
`TransformState` is copied, maybe `const TransformState* getTransformState()`

Reply:
Changed to reference and added assert, there must be a transform state when snapshot is taken.